app/src/server/utils/convertFunctionCalls.ts

```diff
@@ -0,0 +1,96 @@
+import type { ChatCompletionCreateParamsBase } from "openai/resources/chat/completions";
+
+export const convertFunctionCall = (
+  functionCall: ChatCompletionCreateParamsBase["function_call"] | null,
+): ChatCompletionCreateParamsBase["tool_choice"] => {
+  if (!functionCall) {
+    return undefined;
+  } else if (functionCall === "auto" || functionCall === "none") {
+    return functionCall;
+  } else if (functionCall.name) {
+    return {
+      type: "function",
+      function: {
+        name: functionCall.name,
+      },
+    };
+  }
+};
+
+export const convertFunctions = (
+  functions: ChatCompletionCreateParamsBase["functions"] | null,
+): ChatCompletionCreateParamsBase["tools"] => {
+  if (!functions) return undefined;
+  return functions.map((func) => ({
+    type: "function",
+    function: {
+      name: func.name,
+      description: func.description,
+      parameters: func.parameters,
+    },
+  }));
+};
+
+export const convertMessages = (
+  messages: ChatCompletionCreateParamsBase["messages"],
+): ChatCompletionCreateParamsBase["messages"] =>
+  messages ? messages.map(convertFunctionMessageToToolCall) : [];
+
+export const convertFunctionMessageToToolCall = (
+  message: ChatCompletionCreateParamsBase["messages"][0],
+) => {
+  switch (message.role) {
+    case "system":
+    case "user":
+    case "tool":
+      return message;
+    case "assistant":
+      if (message.tool_calls?.length || !message.function_call) {
+        return message;
+      } else {
+        return {
+          ...message,
+          tool_calls: [
+            {
+              id: "",
+              type: "function" as const,
+              function: message.function_call,
+            },
+          ],
+        };
+      }
+    case "function":
```

Comment:
I clearly need to read the new docs, but what's the difference between a `role: "assistant"` message that includes "tool_calls" field and a `role: "tool"`/`role: "function"` message?

Reply:
https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages